toolbox/src/DiscreteFilter.cpp

```diff
@@ -0,0 +1,261 @@
+            error->message = ClassName + " Failed to parse string parameters.";
+        }
+        return false;
+    }
+
+    // Convert the strings from the Matlab syntax ("[1.0 2 3.33]") to yarp::sig::Vector
+    stringToYarpVector(num_coeff_str, num);
+    stringToYarpVector(den_coeff_str, den);
+
+    // Create the filter object
+    // ========================
+
+    // Generic
+    // -------
+    if (filter_type == "Generic") {
+        if (num->length() == 0 || den->length() == 0) {
+            if (error) {
+                error->message = ClassName + " (Generic) Wrong coefficients size";
+            }
+            return 1;
+        }
+        filter = new Filter(*num, *den);
+    }
+    // FirstOrderLowPassFilter
+    // -----------------------
+    else if (filter_type == "FirstOrderLowPassFilter") {
+        if (firstOrderLowPassFilter_fc.floatData() == 0
+            || firstOrderLowPassFilter_ts.floatData() == 0) {
+            if (error) {
+                error->message = ClassName
+                                 + " (FirstOrderLowPassFilter) You need to "
+                                   "specify Fc and Ts";
+            }
+            return false;
+        }
+        filter = new FirstOrderLowPassFilter(firstOrderLowPassFilter_fc.floatData(),
+                                             firstOrderLowPassFilter_ts.floatData());
+    }
+    // MedianFilter
+    // ------------
+    else if (filter_type == "MedianFilter") {
+        if (medianFilter_order.int32Data() == 0) {
+            if (error) {
+                error->message = ClassName
+                                 + " (MedianFilter) You need to specify the "
+                                   "filter order.";
+            }
+            return false;
+        }
+        filter = new MedianFilter(medianFilter_order.int32Data());
+    }
+    else {
+        if (error) error->message = ClassName + " Filter type not recognized.";
+        return false;
+    }
+
+    return true;
+}
+
+bool DiscreteFilter::terminate(BlockInformation* blockInfo, wbt::Error* error)
+{
+    // Deallocate all the memory
+    // -------------------------
+
+    if (filter) {
+        delete filter;
+        filter = nullptr;
+    }
+
+    if (num) {
+        delete num;
+        num = nullptr;
+    }
+
+    if (den) {
+        delete den;
+        den = nullptr;
+    }
+
+    if (inputSignalVector) {
+        delete inputSignalVector;
+        inputSignalVector = nullptr;
+    }
+
+    return true;
+}
+
+bool DiscreteFilter::output(BlockInformation* blockInfo, wbt::Error* error)
+{
+    if (filter == nullptr) return false;
+
+    // Get the input and output signals
+    Signal inputSignal  = blockInfo->getInputPortSignal(INPUT_IDX_SIGNAL);
+    Signal outputSignal = blockInfo->getOutputPortSignal(OUTPUT_IDX_SIGNAL);
+
+    unsigned inputSignalWidth = blockInfo->getInputPortWidth(INPUT_IDX_SIGNAL);
+
+    // Allocate the memory for the input data
+    if (firstRun) {
```

Comment:
In `initialize()` it returns the correct width. I'll add a private member for that and move the initialization into the `initializeInitialConditions` callback that I didn't know existed.